src/Servers/Kestrel/Core/ref/Microsoft.AspNetCore.Server.Kestrel.Core.netcoreapp.cs

```diff
@@ -137,6 +137,7 @@ public partial class KestrelServerOptions
         public bool DisableStringReuse { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public bool EnableAltSvc { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } [System.Runtime.CompilerServices.CompilerGeneratedAttribute] set { } }
         public Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerLimits Limits { [System.Runtime.CompilerServices.CompilerGeneratedAttribute] get { throw null; } }
+        public System.Func<string, System.Text.Encoding> RequestHeaderEncodingSelector { get { throw null; } set { } }
```

Comment:
I have some thoughts about this contract:
- Why string and not `ReadOnlySpan<byte>`?
- Why an Encoding? Why not a delegate? ReadOnlySpan\<byte\> -> string

Reply:
> Why string and not `ReadOnlySpan<byte>`

For all known headers, we can and do pass in HeaderNames.{HeaderName} which you can do a quick reference comparison on. The CanSpecifyEncodingBasedOnHeaderName test verifies this works.

If we passed in `ReadOnlySpan<byte>` instead of string, and the callback wanted a string it would have to decode the name itself. For known headers, Kestrel's decoding is very optimized. And for unknown headers Kestrel already greedily decodes the name, so I seen no reason not to pass that in.

> Why an Encoding? Why not a delegate? ReadOnlySpan<byte> -> string

@stephentoub suggested it, and I agreed Encoding is better. Here is the reasoning:

> I'm thinking that would enable:
>
> - more flexibility in the future, given the various APIs on Encoding
> - the ability to cache an Encoding to use for a given header, which could in turn enable caching opportunities for a given input sequence
> - potential devirtualization opportunities
> Just a thought. Not sure if this has progressed since the original strawman.

https://github.com/dotnet/aspnetcore/issues/17400#issuecomment-634333914

Definitely warrants further discussion though.

Reply:
Could this be a dictionary? It kinda behaves like one.

Reply:
For the two common scenarios, always UTF-8 and always latin1, having a dictionary entry for each known header would super be wasteful. It also wouldn't allow you to easily set the default encoding for unknown headers.

You could theoretically implement `IDictionary<string, Encoding?>` yourself to get the same flexibility, but that's way more complicated than a single callback.